cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/plugin/PluginService.java

```diff
@@ -0,0 +1,221 @@
+  private PluginInstantiator pluginInstantiator;
+
+  public PluginService(ArtifactRepository artifactRepository, File tmpDir, CConfiguration cConf) {
+    this.artifactRepository = artifactRepository;
+    this.stageDir = DirUtils.createTempDir(tmpDir);
+    this.cConf = cConf;
+    this.artifactClassLoaderFactory = new ArtifactClassLoaderFactory(cConf, tmpDir);
+  }
+
+  /**
+   * Given plugin artifact, find the parent artifact of plugin, after creating classloader with parent artifact
+   * and plugin, we invoke the plugin method in the plugin identified by type and name and return the response.
+   * @param namespace namespace
+   * @param artifactId plugin artifact id
+   * @param pluginType type of the plugin
+   * @param pluginName name of the plugin
+   * @param methodName name of the method
+   * @return {@link PluginEndpoint}
+   * @throws IOException
+   * @throws NotFoundException
+   * @throws ClassNotFoundException
+   */
+  public PluginEndpoint getPluginEndpoint(Id.Namespace namespace,
+                                          Id.Artifact artifactId, String pluginType,
+                                          String pluginName, String methodName)
+    throws IOException, NotFoundException, ClassNotFoundException {
+
+    ArtifactDetail artifactDetail = artifactRepository.getArtifact(artifactId);
+    return getPluginEndpoint(namespace, artifactDetail, pluginType, pluginName,
+                             pickParentArtifact(artifactDetail, artifactId), methodName);
+  }
+
+  private ArtifactDescriptor pickParentArtifact(ArtifactDetail artifactDetail, Id.Artifact artifact)
+    throws ArtifactNotFoundException, IOException {
+
+    // get parent artifacts
+    Set<ArtifactRange> parentArtifactRanges = artifactDetail.getMeta().getUsableBy();
+    if (parentArtifactRanges.isEmpty()) {
+      throw new ArtifactNotFoundException(artifact);
+    }
+
+    // just pick the first parent artifact from the set.
+    ArtifactRange parentArtifactRange = parentArtifactRanges.iterator().next();
+
+    List<ArtifactDetail> artifactDetails = artifactRepository.getArtifacts(parentArtifactRange);
+    if (artifactDetails.isEmpty()) {
+      // should not happen
+      throw new ArtifactNotFoundException(artifact);
+    }
+
+    // return the first one from the artifact details list.
+    return artifactDetails.get(0).getDescriptor();
+  }
+
+  private PluginEndpoint getPluginEndpoint(Id.Namespace namespace, ArtifactDetail artifactDetail, String pluginType,
+                                           String pluginName, ArtifactDescriptor parentArtifactDescriptor,
+                                           String methodName)
+    throws NotFoundException, IOException, ClassNotFoundException {
+
+    Id.Artifact artifactId = Id.Artifact.from(namespace, artifactDetail.getDescriptor().getArtifactId());
+    PluginClass pluginClass =
+      artifactRepository.findPlugin(namespace, artifactId, pluginType, pluginName, new PluginSelector()).getValue();
```

Comment:
sorry, I was confused before.  It was correct earlier when you looked inside ArtifactDetail for the PluginClass, since findPlugin will involve IO.